### PR TITLE
Port FSTSyncEngine to C++ SyncEngine 2/2

### PR DESCRIPTION
**Pull request by wu-hui**
Removed FSTSyncEngine.

### CODE REVIEW

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm [hunk 1]
```diff
@@ -199,16 +201,14 @@ - (instancetype)initWithPersistence:(id<FSTPersistence>)persistence
                                                  std::make_shared<EmptyCredentialsProvider>());
     _remoteStore = absl::make_unique<RemoteStore>(
         _localStore, _datastore, _workerQueue, [self](OnlineState onlineState) {
-          [self.syncEngine applyChangedOnlineState:onlineState];
-          _eventManager->HandleOnlineStateChange(onlineState);
+          _syncEngine->HandleOnlineStateChange(onlineState);
+          // _eventManager->HandleOnlineStateChange(onlineState);
```
Comment (var-const) [inline Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm]: Commented-out code.

File: Firestore/core/src/firebase/firestore/core/firestore_client.h [hunk 2]
```diff
@@ -187,11 +183,11 @@ class FirestoreClient : public std::enable_shared_from_this<FirestoreClient> {
...
 
-  id<FSTPersistence> persistence_;
-  FSTLocalStore* local_store_;
+  _Nonnull id<FSTPersistence> persistence_;
+  FSTLocalStore* _Nonnull local_store_;
   std::unique_ptr<remote::RemoteStore> remote_store_;
-  FSTSyncEngine* sync_engine_;
-  util::DelayedConstructor<EventManager> event_manager_;
+  std::unique_ptr<SyncEngine> sync_engine_;
+  std::unique_ptr<EventManager> event_manager_;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/firestore_client.h]: What's the reason for this change?

File: Firestore/core/src/firebase/firestore/core/firestore_client.h [hunk 3]
```diff
@@ -17,10 +17,6 @@
 #ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_FIRESTORE_CLIENT_H_
 #define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_FIRESTORE_CLIENT_H_
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/firestore_client.h]: I think this is still true, no? This class `import`s `<Foundation>` and has some Objective-C pointers.

File: Firestore/Example/Tests/Integration/FSTDatastoreTests.mm [hunk 4]
```diff
@@ -153,6 +154,48 @@ - (void)rejectListenWithTargetID:(const TargetId)targetID error:(NSError *)error
...
+
+  void ExpectWriteEvent(NSString *description) {
+    [underlying_capture_ expectWriteEventWithDescription:description];
+  }
+
+  void ExpectListenEvent(NSString *description) {
+    [underlying_capture_ expectListenEventWithDescription:description];
+  }
+
+  void ApplyRemoteEvent(const remote::RemoteEvent &remote_event) override {
```
Comment (var-const) [inline Firestore/Example/Tests/Integration/FSTDatastoreTests.mm]: Nit: looks like you can just use `RemoteEvent` (there's a `using` declaration above). Applies throughout this file.

File: Firestore/core/src/firebase/firestore/core/firestore_client.mm [hunk 5]
```diff
@@ -427,15 +425,13 @@ QuerySnapshot result(query.firestore(), query.query(), std::move(snapshot),
...
-                if (callback) {
-                  shared_this->user_executor()->Execute(
-                      [=] { callback(Status::FromNSError(error)); });
-                }
-              }];
+      shared_this->sync_engine_->WriteMutations(
+          std::move(mutations), [callback, shared_this](Status error) {
+            // Dispatch the result back onto the user dispatch queue.
+            if (callback) {
+              shared_this->user_executor()->Execute([=] { callback(error); });
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/firestore_client.mm]: Nit: move `error`.

File: Firestore/core/src/firebase/firestore/core/sync_engine.h [hunk 6]
```diff
@@ -47,6 +46,44 @@ namespace firebase {
...
+  virtual ~QueryEventCallback() = default;
+};
+
+/**
+ * Interface implemented by `SyncEngine` receive requests from `EventManager`.
+ */
+class QueryEventSource {
+ public:
+  virtual void SetCallback(QueryEventCallback* callback) = 0;
+  /**
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/sync_engine.h]: Nit: please add an empty line.

File: Firestore/core/src/firebase/firestore/core/sync_engine.h [hunk 7]
```diff
@@ -47,6 +46,44 @@ namespace firebase {
...
+   * Initiates a new listen. The FSTLocalStore will be queried for initial data
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/sync_engine.h]: Nit: like in the previous PR, I'd suggest using just `LocalStore`.

File: Firestore/core/src/firebase/firestore/core/sync_engine.h [hunk 8]
```diff
@@ -47,6 +46,44 @@ namespace firebase {
...
+   * snapshots and/or listen errors.
+   *
+   * @return the target ID assigned to the query.
+   */
+  virtual model::TargetId Listen(Query query) = 0;
+
+  /** Stops listening to a query previously listened to via `Listen`. */
+  virtual void StopListening(const Query& query) = 0;
+
+  virtual ~QueryEventSource() = default;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/sync_engine.h]: Destructor should be declared first within this class, according to the [style guide](https://google.github.io/styleguide/cppguide.html#Declaration_Order):
> _...generally prefer the following order: types (including typedef, using, and nested structs and classes), constants, factory functions, constructors, assignment operators, __destructor, all other methods,__ data members._

Applies to the other interface as well.

File: Firestore/core/src/firebase/firestore/core/sync_engine.h [hunk 9]
```diff
@@ -47,6 +46,44 @@ namespace firebase {
 namespace firestore {
 namespace core {
 
+/**
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/sync_engine.h]: What's the reason to rename this from `SyncEngineCallback`?

File: Firestore/core/src/firebase/firestore/core/sync_engine.h [hunk 10]
```diff
@@ -47,6 +46,44 @@ namespace firebase {
...
+  /** Handles the failure of a query. */
+  virtual void OnError(const core::Query& query, util::Status error) = 0;
+
+  virtual ~QueryEventCallback() = default;
+};
+
+/**
+ * Interface implemented by `SyncEngine` receive requests from `EventManager`.
+ */
+class QueryEventSource {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/sync_engine.h]: Sorry if we discussed this before and I forgot: what's the reason for this decoupling? It doesn't exist on other platforms.

File: Firestore/core/src/firebase/firestore/core/sync_engine.h [hunk 11]
```diff
@@ -47,6 +46,44 @@ namespace firebase {
...
+  /** Handles new view snapshots. */
+  virtual void OnViewSnapshots(std::vector<core::ViewSnapshot>&& snapshots) = 0;
+  /** Handles the failure of a query. */
+  virtual void OnError(const core::Query& query, util::Status error) = 0;
+
+  virtual ~QueryEventCallback() = default;
+};
+
+/**
+ * Interface implemented by `SyncEngine` receive requests from `EventManager`.
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/sync_engine.h]: Nit: "__to__ receive requests" or similar.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+#include "Firestore/core/src/firebase/firestore/core/view_snapshot.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_set.h"
+#include "Firestore/core/src/firebase/firestore/model/types.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/test/firebase/firestore/testutil/testutil.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using firebase::firestore::core::EventListener;
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: Done.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+  MOCK_METHOD1(Listen, model::TargetId(core::Query));
+  MOCK_METHOD1(StopListening, void(const core::Query&));
+};
+
+TEST(EventManagerTest, HandlesManyListnersPerQuery) {
+  core::Query query = Query("foo/bar");
+  auto listener1 = NoopQueryListener(query);
+  auto listener2 = NoopQueryListener(query);
+
+  MockEventSource mock_event_source;
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: Done.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+  EXPECT_CALL(mock_event_source, SetCallback(_));
+  EventManager event_manager(&mock_event_source);
+
+  EXPECT_CALL(mock_event_source, Listen(query));
+  event_manager.AddQueryListener(listener1);
+
+  EXPECT_CALL(mock_event_source, Listen(_)).Times(0);
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: Yes.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+                      DocumentKeySet{},
+                      false,
+                      /*sync_state_changed=*/true,
+                      false};
+}
+
+TEST(EventManagerTest, NotifiesListenersInTheRightOrder) {
+  core::Query query1 = Query("foo/bar");
+  core::Query query2 = Query("bar/baz");
+  std::vector<std::string> eventOrder;
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: Done.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+  auto listener1 =
+      QueryListener::Create(query1, [&eventOrder](StatusOr<ViewSnapshot>) {
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: Done.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+      events.push_back(online_state);
+    }
+
+    std::vector<OnlineState> events;
+  };
+
+  auto fake_listener = std::make_shared<FakeQueryListener>(query);
+
+  MockEventSource mock_event_source;
+  EventManager event_manager(&mock_event_source);
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: See above.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+        eventOrder.push_back("listener3");
+      });
+
+  MockEventSource mock_event_source;
+  EventManager event_manager(&mock_event_source);
+
+  event_manager.AddQueryListener(listener1);
+  event_manager.AddQueryListener(listener2);
+  event_manager.AddQueryListener(listener3);
+
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: I removed these checks intentionally, and the one below, because the behaviour is already tested in `HandlesManyListnersPerQuery`. Adding too many checks distracts reader's attention from the real purpose of this particular case, without adding too much value.

I did what you suggested because I realized `HandlesManyListnersPerQuery` is not testing exactly the same thing, so I'm OK to add them, though I still think checking `event_order` in the last is sufficient for this case.

[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: This is reasonable, but keep in mind that any deviation from other SDKs has an associated cost (anyone modifying those tests in the future across platforms would have to stop and think "Why are these tests doing different things? Which one is correct?"). Changing other SDKs to match is additional work as well. Sometimes, this means that for really minor things, leaving them as they are is the lesser evil.

If you feel strongly about this change, please either add a `PORTING NOTE` comment or backport this change to other platforms as well. I'd personally suggest just following the other SDKs in this particular case, because the benefits of the change seem small, but that's not something I feel super strong about.

File: Firestore/core/src/firebase/firestore/remote/remote_store.h [hunk 19]
```diff
@@ -45,69 +41,62 @@
...
+class RemoteStoreCallback {
+ public:
+  /**
+   * Applies a remote event to the sync engine, notifying any views of the
+   * changes, and releasing any pending mutation batches that would become
+   * visible because of the snapshot version the remote event contains.
+   */
+  virtual void ApplyRemoteEvent(const RemoteEvent& remote_event) = 0;
+  /**
+   * Handles rejection of listen for the given targetId. This can be triggered
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/remote/remote_store.h]: Done.

File: Firestore/core/src/firebase/firestore/core/sync_engine.h (hunk 10 above)
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/core/sync_engine.h]: Done.

File: Firestore/core/src/firebase/firestore/core/sync_engine.h (hunk 9 above)
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/core/sync_engine.h]: Reasonable..I'll just name it back.

File: Firestore/core/test/firebase/firestore/core/event_manager_test.mm (hunk 18 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/firebase/firestore/core/event_manager_test.mm]: Agree in general. 

The current state is already in sync with other platform actually, Android/Web don't check if `setCallback` is called.

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm (hunk 1 above)
Comment (var-const) [reply Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm]: Discussed offline -- it's redundant because the line above calls that already. Great catch; please port this to other platforms as well.
